### PR TITLE
RODS part 2

### PR DESCRIPTION
**Pull request by kvark**
**Connections**
Unblocks https://github.com/gfx-rs/wgpu-rs/issues/359
Is a follow-up to RODS part 1 - #685

**Description**
There is a few things in here.

Thing 1: Stronger assertions on the load/store ops of the depth & stencil.

### Thing 2: rewritten tracking of render attachments
Background: from the usage tracker point of view, each subresource can be in either "extend" mode, where it accumulates a single combined usage, or in the "replace" mode, where it goes from one usage to another, producing the relevant transitions on the way.

The problem turned out to come from the fact that the render pass attachments were always tracked in "replace" mode. This is needed because their track don't have a single state: render pass itself encodes a transition of attachments. However, it also means that there was no way to "extend" the usage in RODS scenarios...

So I could see two ways to address this:
  - re-achitecture the tracking a bit more in general, representing it as a sequence of merges.
  - introduce the "prepend()" tracking operator that's *only* used for render pass attachments

I opted for the latter as it seems much less intrusive. The render pass attachments accumulate their usage like everything else in the "extend mode". But right before we are inserting the transitions (between the active command buffer and the pass), we turn the tracking of the attachments from "extend" into "replace" mode by installing the "first" usage according to what we expect it to be.

### Thing 3: missing API for RODS bind groups
The original RODS design missed a problem with Vulkan image layouts. When creating a bind group, one has to specify what layout the image will be in. We always used `ShaderReadOnlyOptimal` until now for texture views. However, in RODS scenarios this has to be `DepthStencilReadOnlyOptimal`. Luckily, it's compatible with sampling from the shader, but we still need to know about this when creating the bind group.

**Testing**
Tested on the modified water example provided in https://github.com/gfx-rs/wgpu-rs/issues/359#issuecomment-642167269 

Added a few tests to the buffer implementation of the new `prepend()` operator.

### CODE REVIEW

**Comment by cwfitzgerald** [comment]
Testing on vulkan...

**Comment by cwfitzgerald** [comment]
Using the latest commit in this branch with the example given in the issue still gives me tons of validation issues:

<details>
<summary>Errors</summary>

```
[2020-06-11T03:58:27Z ERROR gfx_backend_vulkan] 
    VALIDATION [UNASSIGNED-CoreValidation-Shader-FeatureNotEnabled (0)] : Shader requires VkPhysicalDeviceFeatures::shaderClipDistance but is not enabled on the device
    object info: (type: UNKNOWN, hndl: 0)
    
[2020-06-11T03:58:27Z ERROR gfx_backend_vulkan] 
    VALIDATION [VUID-VkDescriptorImageInfo-imageLayout-00344 (0)] : vkCmdDraw(): Cannot use VkImage 0x1f000000001f[] (layer=0 mip=0) with specific layout VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL that doesn't match the previous known layout VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL. The Vulkan spec states: imageLayout must match the actual VkImageLayout of each subresource accessible from imageView at the time this descriptor is accessed as defined by the image layout matching rules (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkDescriptorImageInfo-imageLayout-00344)
    object info: (type: COMMAND_BUFFER, hndl: 93900869250320)
    
[2020-06-11T03:58:27Z ERROR gfx_backend_vulkan] 
    VALIDATION [UNASSIGNED-CoreValidation-DrawState-DescriptorSetNotUpdated (0)] : VkDescriptorSet 0x2c000000002c[] bound as set #0 encountered the following validation error at vkCmdDraw() time: Image layout specified at vkUpdateDescriptorSet* or vkCmdPushDescriptorSet* time doesn't match actual image layout at time descriptor is used. See previous error callback for specific details.
    object info: (type: DESCRIPTOR_SET, hndl: 48378511622188)
    
[2020-06-11T03:58:27Z ERROR gfx_backend_vulkan] 
    VALIDATION [VUID-VkDescriptorImageInfo-imageLayout-00344 (0)] : vkCmdDraw(): Cannot use VkImage 0x1f000000001f[] (layer=0 mip=0) with specific layout VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL that doesn't match the previous known layout VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL. The Vulkan spec states: imageLayout must match the actual VkImageLayout of each subresource accessible from imageView at the time this descriptor is accessed as defined by the image layout matching rules (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkDescriptorImageInfo-imageLayout-00344)
    object info: (type: COMMAND_BUFFER, hndl: 93900867357248)
    
[2020-06-11T03:58:27Z ERROR gfx_backend_vulkan] 
    VALIDATION [UNASSIGNED-CoreValidation-DrawState-DescriptorSetNotUpdated (0)] : VkDescriptorSet 0x2c000000002c[] bound as set #0 encountered the following validation error at vkCmdDraw() time: Image layout specified at vkUpdateDescriptorSet* or vkCmdPushDescriptorSet* time doesn't match actual image layout at time descriptor is used. See previous error callback for specific details.
    object info: (type: DESCRIPTOR_SET, hndl: 48378511622188)
    
[2020-06-11T03:58:27Z ERROR gfx_backend_vulkan] 
    VALIDATION [UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout (0)] : Submitted command buffer expects VkImage 0x1f000000001f[] (subresource: aspectMask 0x2 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_GENERAL--instead, current layout is VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL.
    object info: (type: COMMAND_BUFFER, hndl: 93900867332320)
```
and on and on like that

</details>

**Comment by kvark** [comment]
Thank you for testing @cwfitzgerald ! This is a bit worrying, I think it shows that we are missing a bit in terms of the API for RODS. We need to know if a particular `BindGroupLayoutEntry` will be used with read-only depth, in order to set the proper image layout associated with it. I'll suggest that to the group.

**Comment by OptimisticPeach** [comment]
If it matters, the errors mention user clipping planes, which aren't supported on all hardware and in some cases must be enabled. If that may be a (separate) issue, it may be best to remove that from the shader for the time being. Its usage is in the terrain vertex shader.

File: wgpu-core/src/command/render.rs [hunk 1]
```diff
@@ -427,6 +446,15 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         // instead of the special read-only one, which would be `None`.
         let mut is_ds_read_only = false;
 
+        struct OutputAttachment<'a> {
+            texture_id: &'a Stored<id::TextureId>,
+            range: &'a hal::image::SubresourceRange,
+            previous_use: Option<TextureUse>,
+            new_use: TextureUse,
+        }
+        const MAX_TOTAL_ATTACHMENTS: usize = MAX_COLOR_TARGETS + MAX_COLOR_TARGETS + 1;
```
Comment (cwfitzgerald) [inline wgpu-core/src/command/render.rs:455]: Either this is a typo or should be `2 * MAX_COLOR_TARGETS + 1`

Comment (kvark) [reply wgpu-core/src/command/render.rs:455]: That's intentional :)
I don't mind changing it to what you suggest

Comment (cwfitzgerald) [reply wgpu-core/src/command/render.rs]: Alright :) It just looked like it could have been a typo for COLOR + DEPTH_STENCIL or whatever

**Comment by kvark** [comment]
Review note is addressed. Thank you for having a look!
bors r=cwfitzgerald

**Pull request merged**